src/mmap_xen.rs

```diff
@@ -0,0 +1,607 @@
+        }
+    }
+
+    fn unmap_ioctl(&self, count: u32, index: u64) -> Result<()> {
+        let unmap = GntDevUnmapGrantRef::new(index, count);
+
+        // SAFETY: This is safe because the ioctl guarantees to not access memory beyond unmap.
+        let ret = unsafe { ioctl_with_ref(self, ioctl_gntdev_unmap_grant_ref(), &unmap) };
+
+        if ret == 0 {
+            Ok(())
+        } else {
+            Err(Error::Mmap(io::Error::last_os_error()))
+        }
+    }
+
+    fn mmap_internal(&self, addr: GuestAddress, size: usize, prot: i32) -> Result<(*mut u8, u64)> {
+        let (count, size) = pages(size);
+        let index = self.mmap_ioctl(addr, count)?;
+        let addr = mmap_unix::mmap_unix(size, prot, self.flags, self.fd, index)?;
+
+        Ok((addr, index))
+    }
+
+    fn munmap_internal(&self, unmap_addr: *mut u8, size: usize, index: u64) {
+        let (count, size) = pages(size);
+
+        // SAFETY: This is safe because we mapped the addr earlier.
+        unsafe {
+            libc::munmap(unmap_addr as *mut c_void, size);
+        }
+
+        self.unmap_ioctl(count as u32, index).unwrap();
+    }
+}
+
+impl MmapXenTrait for MmapXenGrant {
+    fn mmap_in_advance(&self) -> bool {
+        self.mmap_in_advance
+    }
+
+    // Maps the memory if mmap_in_advance() is set.
+    fn mmap(
+        &mut self,
+        size: usize,
+        prot: i32,
+        flags: i32,
+        file_offset: &Option<FileOffset>,
+    ) -> Result<*mut u8> {
+        let (fd, _) = validate_file(file_offset)?;
+
+        // Save fd and flags for later use.
+        self.fd = fd;
+        self.flags = flags;
+
+        // Region can't be mapped in advance, partial mapping will be done later via
+        // `MmapXenSlice`.
+        if !self.mmap_in_advance() {
+            return Ok(null_mut());
+        }
+
+        (self.addr, self.index) = self.mmap_internal(self.guest_base, size, prot)?;
+        self.size = size;
+        Ok(self.addr)
+    }
+
+    // Unmaps the memory in architecture / platform dependent way.
+    fn munmap(&self) {
+        if self.mmap_in_advance() && self.size != 0 {
+            self.munmap_internal(self.addr, self.size, self.index);
+        }
+    }
+
+    // Maps a slice out of the entire region.
+    fn mmap_slice(&self, addr: *const u8, is_read: bool, len: usize) -> Result<MmapXenSlice> {
+        assert!(!self.mmap_in_advance());
+        MmapXenSlice::new_with(self.clone(), addr as usize, is_read, len)
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct MmapXenSlice {
+    grant: Option<MmapXenGrant>,
+    unmap_addr: *mut u8,
+    addr: *mut u8,
+    size: usize,
+    index: u64,
+}
+
+impl MmapXenSlice {
+    pub fn new(addr: *mut u8) -> Self {
+        Self {
+            grant: None,
+            unmap_addr: null_mut(),
+            addr,
+            size: 0,
+            index: 0,
+        }
+    }
+
+    fn new_with(grant: MmapXenGrant, offset: usize, is_read: bool, size: usize) -> Result<Self> {
```

Comment:
While doing my hacks, I realized that this `is_read` may be easy to interpret incorrectly. Maybe an enum of `READ_ONLY`, `WRITE_ONLY` would make things more clear?

Reply:
I have passed `prot` directly now, which can be libc::PROT_READ/WRITE.

Reply:
Ok. Though that may tempt people to try passing `READ | WRITE` (at least I tried when hacking the virtio-fs test together), but that did not seem to work since the mapping needs to be either readable OR writable apparently?

Reply:
Yeah, I was just doing it as a quick and dirty hack. But mmap blew up, so I figured it out quickly.